File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/DoNotHardcodeCredentials.SecureString.vb
Imports System
Imports System.Security

Namespace Tests.Diagnostics
    Class Issue_7323
        Public Sub Noncompliant()
            Using securePwd As SecureString = New SecureString()
                For i As Integer = 0 To "AP@ssw0rd".Length - 1
                    securePwd.AppendChar("AP@ssw0rd"(i)) ' Noncompliant {{Please review this hard-coded password.}}
                Next
            End Using
        End Sub

        Public Sub Compliant(ByVal password As String)
            Using securePwd As SecureString = New SecureString()
                For i As Integer = 0 To password.Length - 1
                    securePwd.AppendChar(password(i))
                Next
            End Using
        End Sub
    End Class

    Class Tests
        Public Sub Characters()
            Using securePwd = New SecureString()
                securePwd.AppendChar("P"c) ' Noncompliant
                securePwd.AppendChar("a"c) ' Noncompliant
                securePwd.AppendChar("s"c) ' Noncompliant
                securePwd.AppendChar("s"c) ' Noncompliant
                Dim w = "w"c
                securePwd.AppendChar(w)    ' FN
            End Using
        End Sub

        Public Sub Constant()
            Const keyword As String = "AP@ssw0rd"
            Using securePwd As SecureString = New SecureString()

                For i As Integer = 0 To keyword.Length - 1
                    securePwd.AppendChar(keyword(i)) ' Noncompliant
                Next
            End Using
        End Sub

        Public Sub ForEachOverString()
            Using securePwd As SecureString = New SecureString()
                For Each c In "AP@ssw0rd"
                    securePwd.AppendChar(c) ' Noncompliant
                Next
            End Using
        End Sub

        Public Sub FromUnmodifiedVariable_ForLoop()
            Dim keyword = "AP@ssw0rd"
            Using securePwd As SecureString = New SecureString()

                For i As Integer = 0 To keyword.Length - 1
                    securePwd.AppendChar(keyword(i)) ' Noncompliant
                Next
            End Using
        End Sub

        Public Sub FromUnmodifiedVariable_ForEachLoop()
            Dim keyword = "AP@ssw0rd"
            Using securePwd As SecureString = New SecureString()

                For Each c In keyword
                    securePwd.AppendChar(c) ' Noncompliant
                Next
            End Using
        End Sub

        Public Sub FromBytes()
            Dim bytes = New Byte() {80, 97, 115, 115, 119, 111, 114, 100}
            Using securePwd As SecureString = New SecureString()
                For i As Integer = 0 To bytes.Length - 1
                    securePwd.AppendChar(Convert.ToChar(bytes(i))) ' FN.
                Next
            End Using
        End Sub

        Public Sub FromCharArray()
            Dim chars = New Char() {"P"c, "a"c, "s"c, "s"c, "w"c, "o"c, "r"c, "d"c}
            Using securePwd As SecureString = New SecureString()
                For i As Integer = 0 To chars.Length - 1
                    securePwd.AppendChar(chars(i)) ' FN.
                Next
            End Using
        End Sub

        Public Sub FromExpression()
            Dim chars = New Char() { "P"c, "a"c, "s"c, "s"c }
            Using securePwd As SecureString = New SecureString()
                For i As Integer = 0 To chars.Length - 1
                    securePwd.AppendChar("a"c + chars(i)) ' FN.
                Next
            End Using
        End Sub
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/DoNotHardcodeCredentials.SecureString.cs
using System;
using System.Security;

namespace Tests.Diagnostics
{
    class Issue_7323
    {
        // https://github.com/SonarSource/sonar-dotnet/issues/7323
        public void Noncompliant()
        {
            using (SecureString securePwd = new SecureString())
            {
                for (int i = 0; i < "AP@ssw0rd".Length; i++)
                {
                    securePwd.AppendChar("AP@ssw0rd"[i]); // Noncompliant {{Please review this hard-coded password.}}
                }

                // Do something with securePwd
            }
        }

        public void Compliant(String password)
        {
            using (SecureString securePwd = new SecureString())
            {
                for (int i = 0; i < password.Length; i++)
                {
                    securePwd.AppendChar(password[i]); // Compliant
                }

                // Do something with securePwd
            }
        }
    }

    class Tests
    {
        public void Characters()
        {
            using (var securePwd = new SecureString())
            {
                securePwd.AppendChar('P'); // Noncompliant
                securePwd.AppendChar('a'); // Noncompliant
                securePwd.AppendChar('s'); // Noncompliant
                securePwd.AppendChar('s'); // Noncompliant
                var w = 'w';
                securePwd.AppendChar(w);   // FN
            }
        }

        public void Constant()
        {
            const string keyword = "AP@ssw0rd";
            using (SecureString securePwd = new SecureString())
            {
                for (int i = 0; i < keyword.Length; i++)
                {
                    securePwd.AppendChar(keyword[i]); // Noncompliant
                }

                // Do something with securePwd
            }
        }

        public void ForEachOverString()
        {
            using (SecureString securePwd = new SecureString())
            {
                foreach (var c in "AP@ssw0rd")
                {
                    securePwd.AppendChar(c); // Noncompliant
                }

                // Do something with securePwd
            }
        }

        public void FromUnmodifiedVariable_ForLoop()
        {
            var keyword = "AP@ssw0rd";
            using (SecureString securePwd = new SecureString())
            {
                for (int i = 0; i < keyword.Length; i++)
                {
                    securePwd.AppendChar(keyword[i]); // Noncompliant
                }
            }
        }

        public void FromUnmodifiedVariable_ForEachLoop()
        {
            var keyword = "AP@ssw0rd";
            using (SecureString securePwd = new SecureString())
            {
                foreach(var c in keyword)
                {
                    securePwd.AppendChar(c); // Noncompliant
                }
            }
        }

        public void FromBytes()
        {
            var bytes = new byte[] { 80, 97, 115, 115, 119, 111, 114, 100 };
            using (SecureString securePwd = new SecureString())
            {
                for (int i = 0; i < bytes.Length; i++)
                {
                    securePwd.AppendChar(Convert.ToChar(bytes[i])); // FN.
                }
            }
        }

        public void FromCharArray()
        {
            var chars = new char[] { 'P', 'a', 's', 's', 'w', 'o', 'r', 'd' };
            using (SecureString securePwd = new SecureString())
            {
                for (int i = 0; i < chars.Length; i++)
                {
                    securePwd.AppendChar(chars[i]); // FN.
                }
            }
        }
    }

    class AppendCharFromOtherType
    {
        void AppendChar(char c) { }

        void Test()
        {
            var other = new AppendCharFromOtherType();
            const string keyword = "AP@ssw0rd";
            for (int i = 0; i < keyword.Length; i++)
            {
                other.AppendChar(keyword[i]); // Compliant. Not SecureString.AppendChar
            }
        }
    }
}
